tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py

```diff
@@ -0,0 +1,304 @@
+        print("Number of clusters used = {}".format(n_clusters))
+        assert "doc_id" not in df.columns
+        df["embeddings"] = ""  # refers to document embeddings
+        df = df.map_partitions(self._get_doc_embeddings, meta=df)
+        # df = dd.from_pandas(pdf)
+
+        df = df.explode("documents")
+        df = df.map_partitions(self.assign_ids)
+        embeddings_dataset = DocumentDataset(df)
+
+        self.clustering_model = ClusteringModel(
+            id_column="doc_id",
+            max_iter=100,
+            n_clusters=n_clusters,
+            clustering_output_dir=self.cluster_output_dir,
+            logger=self.logger_output_dir,
+        )
+        clustered_dataset = self.clustering_model(embeddings_dataset)
+        df_c = clustered_dataset.df
+        df_c = df_c[["documents", "question"]]
+
+        return DocumentDataset(df_c)
+
+    def _get_doc_embeddings(self, p_df: pd.DataFrame):
+
+        if self.model_type == "nvidia":
+            self.client = load_object_on_worker(
+                attr="nim_embedding_model",
+                load_object_function=create_nim_client,
+                load_object_kwargs={"base_url": self.base_url, "api_key": self.api_key},
+            )
+            p_df["embeddings"] = p_df["documents"].map(
+                lambda pgs: [self._get_nim_embedding(t, "passage") for t in pgs]
+            )
+        elif self.model_type == "hf":
+            self.hf_model = load_object_on_worker(
+                attr="hf_embedding_model",
+                load_object_function=create_hf_model,
+                load_object_kwargs={"model_name_or_path": self.model_name},
+            )
+            p_df["embeddings"] = p_df["documents"].map(
+                lambda pgs: [
+                    self._get_hf_embedding(t, self.passage_prefix) for t in pgs
+                ]
+            )
+        return p_df
+
+    def _groupby_question(self, pdf):
+        pdf2 = pdf.groupby("question").agg({"documents": set})
+        pdf2["documents"] = pdf2["documents"].map(lambda x: list(x))
+        del pdf
+        return pdf2
+
+    def __call__(self, dataset: DocumentDataset) -> DocumentDataset:
+
+        df = dataset.df
+        df = df.to_backend("pandas")
+        df = df[["question", "documents"]]
+        df = df.map_partitions(self._groupby_question).reset_index()
+        print("Number partitions in dataset = {}".format(df.npartitions))
+
+        df["neg_doc_scores"] = ""
+        df["neg_doc"] = ""
+        df["doc_embed"] = ""
+        df["query_embed"] = ""
+        df["min_pos_score"] = ""
+
+        with ProgressBar(dt=1):
+            df = df.map_partitions(self._process_partition, meta=df).compute()
+
+        df = df.rename(columns={"documents": "pos_doc"})
+        df = df[["question", "pos_doc", "neg_doc"]]
+
+        return DocumentDataset(dd.from_pandas(df))
+
+    def _process_partition(self, df_p: pd.DataFrame):
+
+        if self.model_type == "nvidia":
```

Comment:
Can we change this model_type name to `nim` or `open_ai client` , model_type="nvidia" is confusing

```suggestion
        if self.model_type == "nim":
```